File: backend/src/main/java/br/com/calciolari/datahub/imports/domain/parser/ParseIssue.java
package br.com.calciolari.datahub.imports.domain.parser;

import java.util.Objects;

/**
 * Structured parser/validation issue. Messages must be sanitized; never embed raw
 * binary snippets.
 */
public record ParseIssue(
		String code,
		IssueSeverity severity,
		IssueStage stage,
		SourceLocator sourceLocator,
		String message
) {
	public ParseIssue {
		Objects.requireNonNull(code, "code");
		Objects.requireNonNull(severity, "severity");
		Objects.requireNonNull(stage, "stage");
		Objects.requireNonNull(message, "message");
		if (sourceLocator == null) {
			sourceLocator = SourceLocator.empty();
		}
	}
}

File: backend/src/Calciolari.DataHub/Imports/Domain/Parser/ParseIssue.cs
namespace Calciolari.DataHub.Imports.Domain.Parser;

/// <summary>
/// Structured parser/validation issue. Messages must be sanitized; never embed raw
/// binary snippets.
/// </summary>
public sealed record ParseIssue(
    string Code,
    IssueSeverity Severity,
    IssueStage Stage,
    SourceLocator SourceLocator,
    string Message)
{
    public static ParseIssue Create(
        string code,
        IssueSeverity severity,
        IssueStage stage,
        SourceLocator? sourceLocator,
        string message)
    {
        ArgumentNullException.ThrowIfNull(code);
        ArgumentNullException.ThrowIfNull(message);
        return new ParseIssue(code, severity, stage, sourceLocator ?? SourceLocator.Empty, message);
    }
}
